ethcore/src/blockchain.rs

```diff
@@ -517,27 +635,57 @@ impl BlockChain {
 				for (index, hash) in route.blocks.iter().skip(route.index).enumerate() {
 					batch.put_extras(&(start_number + index as BlockNumber), hash);
 				}
+
+				// get all blocks that are not part of canon chain (TODO: optimize it to one query)
+				let blooms: Vec<H2048> = route.blocks.iter()
+					.skip(route.index)
+					.map(|hash| self.block(hash).unwrap())
+					.map(|bytes| BlockView::new(&bytes).header_view().log_bloom())
+					.collect();
+
+				// reset blooms chain head
+				modified_blooms = ChainFilter::new(self, self.bloom_indexer.index_size(), self.bloom_indexer.levels())
+					.reset_chain_head(&blooms, start_number as usize, self.best_block_number() as usize);
 			},
 			// route.blocks.len() could be 0 only if inserted block is best block,
 			// and this is not possible at this stage
 			_ => { unreachable!(); }
 		};
 
+		let bloom_hashes = modified_blooms.iter()
+			.map(|(bloom_index, _)| self.bloom_indexer.location(&bloom_index).hash)
+			.collect();
+
+		for (bloom_hash, blocks_blooms) in modified_blooms.into_iter()
+			.fold(HashMap::new(), | mut acc, (bloom_index, bloom) | {
+			{
+				let location = self.bloom_indexer.location(&bloom_index);
+				let mut blocks_blooms = acc
+					.entry(location.hash.clone())
+					.or_insert_with(|| self.blocks_blooms(&location.hash).unwrap_or_else(BlocksBlooms::new));
+				blocks_blooms.blooms[location.index] = bloom;
```

Comment:
That's a good question... there is no proof right now. I've added assertion here: https://github.com/ethcore/parity/commit/4b69b96f9bef717f08c50c46ad7280f4513e7dc1